### PR TITLE
Pass composition locals at the very top level of the Window/Dialog

### PR DESCRIPTION
**Pull request by igordmn**
Currently we pass them not at the top level:
```
composition.setContent {
    ProvideCommonCompositionLocals {
        CompositionLocalProvider(compositionLocalContext) { ... }
    }
}
```
And because of that, compositionLocalContext overrides all locals provided by ProvideCommonCompositionLocals. For example, in this code:
```
application {
  Window {
     Dialog {
     }
  }
}
```
application provides a value for default LocalDensity, and pass it down to Window, and Dialog. But Window and Dialog should provide their own density, as they know on which display they are open.

To fix that, we change order of the providing locals:
```
composition.setContent {
    CompositionLocalProvider(compositionLocalContext) {
       ProvideCommonCompositionLocals { ... }
    }
}
```

Fixes JetBrains/compose-jb#1329
Maybe fixes JetBrains/compose-jb#1376 (need to check)

Partially fixes:
JetBrains/compose-jb#1193
JetBrains/compose-jb#1223
JetBrains/compose-jb#990

### CODE REVIEW

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeLayer.desktop.kt [hunk 1]
```diff
@@ -235,6 +236,12 @@ internal class ComposeLayer {
         isDisposed = true
     }
 
+    var compositionLocalContext: CompositionLocalContext?
+        get() = scene.compositionLocalContext
```
Comment (olonho) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeLayer.desktop.kt]: Maybe https://kotlinlang.org/docs/delegated-properties.html#delegating-to-another-property

File: compose/ui/ui/src/desktopTest/kotlin/androidx/compose/ui/awt/ComposePanelTest.kt [hunk 2]
```diff
@@ -33,7 +33,6 @@ class ComposePanelTest {
                 frame.contentPane.add(composePanel)
                 frame.isUndecorated = true
 
-                frame.isVisible = true
```
Comment (olonho) [inline compose/ui/ui/src/desktopTest/kotlin/androidx/compose/ui/awt/ComposePanelTest.kt:36]: Why this change?

File: compose/ui/ui/src/desktopTest/kotlin/androidx/compose/ui/window/window/WindowTest.kt [hunk 3]
```diff
@@ -466,4 +468,25 @@ class WindowTest {
...
+                CompositionLocalProvider(LocalDensity provides customDensity) {
+                    Window(onCloseRequest = ::exitApplication) {
+                        actualDensity = LocalDensity.current
+                    }
+                }
+            }
+        }
+
+        awaitIdle()
+        assertThat(actualDensity).isNotEqualTo(customDensity)
```
Comment (olonho) [inline compose/ui/ui/src/desktopTest/kotlin/androidx/compose/ui/window/window/WindowTest.kt:489]: If `actualDensity == null` the test will still pass.

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeDialog.desktop.kt [hunk 4]
```diff
@@ -64,6 +65,14 @@ class ComposeDialog(
         content = content
     )
 
+    // TODO(CL) non-experimental new API. we can't remove it when we merge it into AOSP, we can just deprecate it.
```
Comment (olonho) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeDialog.desktop.kt:68]: Cannot it be internal?

File: compose/ui/ui/src/desktopTest/kotlin/androidx/compose/ui/awt/ComposePanelTest.kt (hunk 2 above)
Comment (igordmn) [reply compose/ui/ui/src/desktopTest/kotlin/androidx/compose/ui/awt/ComposePanelTest.kt:36]: `pack` doesn't work as expected immediately after `frame.isVisible=true`. The test was failing sometimes here:
```
assertThat(frame.size).isEqualTo(Dimension(234, 345))
```
the actual size was (2,3).

I don't know if the issue in Skiko, or in Swing, but the purpose of the test isn't checking `pack` immediately after `isVisible` (which is very rare use case)

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeLayer.desktop.kt (hunk 1 above)
Comment (igordmn) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeLayer.desktop.kt]: Wow, I was waiting for this feature for years :). Fixed.

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeDialog.desktop.kt (hunk 4 above)
Comment (igordmn) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeDialog.desktop.kt:68]: No, otherwise third-party developers can't pass composition locals the proper way. I already know, that Aurora uses it.

File: compose/ui/ui/src/desktopTest/kotlin/androidx/compose/ui/window/window/WindowTest.kt (hunk 3 above)
Comment (igordmn) [reply compose/ui/ui/src/desktopTest/kotlin/androidx/compose/ui/window/window/WindowTest.kt:489]: Indeed, thanks for noticing!

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeDialog.desktop.kt (hunk 4 above)
Comment (olonho) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeDialog.desktop.kt:68]: Maybe mark it as experimental at least?

**Review by olonho — APPROVED** [review]

**Review by Rsedaikin — APPROVED** [review]

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeDialog.desktop.kt (hunk 4 above)
Comment (igordmn) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeDialog.desktop.kt:68]: We can deprecate it, if realize it is not we want in the future.

Better to provide backward-compatible API for the libraries
